src/runtimes/node/utils/entry_file.ts

```diff
@@ -46,14 +50,38 @@ export const conflictsWithEntryFile = (
   srcFiles: string[],
   {
     basePath,
-    mainFile,
+    extension,
+    featureFlags,
     filename,
+    mainFile,
   }: {
     basePath: string
+    extension: string
+    featureFlags: FeatureFlags
     filename: string
     mainFile: string
   },
-) => srcFiles.some((srcFile) => isNamedLikeEntryFile(srcFile, { basePath, filename }) && srcFile !== mainFile)
+) => {
+  let hasConflict = false
+
+  srcFiles.forEach((srcFile) => {
+    if (featureFlags.zisi_disallow_new_entry_name && srcFile.includes(ENTRY_FILE_NAME)) {
```

Comment:
Why are you using `includes` here?

Reply:
To catch dirs as well as files that contain the name. It would also catch names that only include `___netlify-lambda-entry` as part of it, but it is simpler, the alternative would be:

```js
basename(srcFile, extname(srcFile)) === ENTRY_FILE_NAME || srcFile.includes(`/${ENTRY_FILE_NAME}/`)
```

Reply:
I see. I find it a bit disconcerting that we're checking so loosely for that string anywhere in the file or directory names, even if that's a very unique string.

But more generally, I worry if we need to do this type of check at all. We're the ones adding the entry file, so we know exactly what is its name and extension. Can we check, at the point of inserting it, whether there's already a file in the zip with that name? At that point, it becomes a simple equality check, which should also be more performant?

Reply:
The problem is that it is not just one file, but multiple that we are forbidding. When AWS searches for the entrypoint this is what it looks for (tested just now):

1. /___netlify-lambda-entry.js
2. /___netlify-lambda-entry.mjs
3. /___netlify-lambda-entry.cjs
4. /___netlify-lambda-entry/index.js
5. /___netlify-lambda-entry/index.mjs
6. /___netlify-lambda-entry/index.cjs

So any user file named like this should not be allowed. Checking when adding the entry file is a little too late for now, because this PR is meant to check if `___netlify-lambda-entry` is already used. And as we currently do not always generate an entry file, we would not always check for these files.

Alternatively we could have a Set with all those relative paths and check if the user files is included?

Reply:
Once we generate the file (lets say `/___netlify-lambda-entry.mjs`) this would change, as the only thing we need to forbid is `/___netlify-lambda-entry.js` (because it has higher prior than its mjs equvivalent) and `/___netlify-lambda-entry.mjs`

Reply:
Ah, I see!

> Alternatively we could have a Set with all those relative paths and check if the user files is included?

That makes sense to me. If you're okay with that, I would suggest this route.

Reply:
Worse, some combinations make Lambda crash I just noticed:
Crashes:
<img width="1780" alt="Screenshot 2023-01-31 at 15 27 08" src="https://user-images.githubusercontent.com/231804/215787833-6a86e44a-4387-4c3f-8f7b-29445b8c8a7e.png">

Rename the folder:
<img width="1767" alt="Screenshot 2023-01-31 at 15 27 28" src="https://user-images.githubusercontent.com/231804/215787909-7c8fa829-43c8-45ac-ae71-7155452dd0f7.png">


I will look into the Set.

Reply:
Disallowing `/___netlify-lambda-entry/` and `___netlify-lambda-entry.(c|m)?js` would look like this:
I am not sure I like this, and also does a lot more per iteration.

```diff
@@ -1,4 +1,4 @@
-import { basename, extname, resolve } from 'path'
+import { basename, extname, relative, resolve } from 'path'
 
 import type { FeatureFlags } from '../../../feature_flags.js'
 import { FunctionBundlingUserError } from '../../../utils/error.js'
@@ -24,6 +24,12 @@
   return `export { handler } from '${importPath}'`
 }
 
+const LAMBDA_ENTRY_FILES = new Set([
+  ENTRY_FILE_NAME + ModuleFileExtension.JS,
+  ENTRY_FILE_NAME + ModuleFileExtension.MJS,
+  ENTRY_FILE_NAME + ModuleFileExtension.CJS,
+])
+
 // They are in the order that AWS Lambda will try to find the entry point
 const POSSIBLE_LAMBDA_ENTRY_EXTENSIONS = [ModuleFileExtension.JS, ModuleFileExtension.MJS, ModuleFileExtension.CJS]
 
@@ -65,7 +71,12 @@
   let hasConflict = false
 
   srcFiles.forEach((srcFile) => {
-    if (featureFlags.zisi_disallow_new_entry_name && srcFile.includes(ENTRY_FILE_NAME)) {
+    const relativeSrcFile = relative(basePath, srcFile)
+
+    if (
+      featureFlags.zisi_disallow_new_entry_name &&
+      (relativeSrcFile.startsWith(`${ENTRY_FILE_NAME}/`) || LAMBDA_ENTRY_FILES.has(relativeSrcFile))
+    ) {
       throw new FunctionBundlingUserError(
         `'${ENTRY_FILE_NAME}' is a reserved word and cannot be used as a file or directory name.`,
         {

```

Reply:
Agreed.